Fix issues with ENV config and partially defined config

jason-o-matic opened the pull request.

jason-o-matic committed:
Fix issue loading ENV based config

It would return a path instead of a config previously.

jason-o-matic committed:
Fix potential issue with partial config

If you specify your own config, but you only specify the keys you care about, you'll get errors beacuse we expect a fully defined config.

jason-o-matic committed:
Fix test since changing config loading

janxious commented:
![wut](http://i3.kym-cdn.com/entries/icons/original/000/022/978/yNlQWRM.jpg)

Can you write a description of what you're trying to make happen? In particular, 4303416 is not something I really want I think.

jason-o-matic commented:
If you define a config file, but you don't specify *every single attribute* then things will explode. This isn't how config files work in general. It's easy to imagine a customer making a config file that only specifies the specific thing they want to change.

The change to config loading makes it so the various config files get merged together with the precedence that you'd expect. This makes sure that all config values are defined when we go to look them up later.

I guess we could change things around so we dynamically load fallback config files if a specific config value isn't found, but that seems more complicated.

I think providing sane defaults on a per config attribute basis is a good idea in general. I'm open to other approaches than what's in this PR if you have ideas.

janxious commented:
Well, here's the thing: I think you've taken something straightforward and made it more complex than it needs to or should be. We talked **at length** about configs on multiple occasions and came to the conclusion that what was already in the gem was a good/reasonable approach. What changed your mind?

-----

> If you define a config file, but you don't specify every single attribute then things will explode. This isn't how config files work in general.

Yes, and it makes configuring software with which you're not intimately familiar very difficult.

> It's easy to imagine a customer making a config file that only specifies the specific thing they want to change.

If all our messaging about configs has steps contra to that, why would they?

> The change to config loading makes it so the various config files get merged together with the precedence that you'd expect. 

I disagree. Mentally understanding a three way merge should not be necessary for anyone who puts themselves in that situation. If the env version is defined, it should *just* use that over the application one.

> I think providing sane defaults on a per config attribute basis is a good idea in general.

We do, and they're codified in a file, which is easy to find and copy to somewhere in your application space.

-----

> I'm open to other approaches than what's in this PR if you have ideas.

The things we agreed upon that would be better than this but that we were punting on were:

1. check for the presence of the keys we expect in user-defined configs and warn at startup if any are missing, with helpful information
2. add a thor/rake task to copy over/merge settings from the gem into `application/config`

As this PR does neither of those, I am 👎 without more information.

janxious commented:
@jason-o-matic and I talked about this in person. We are going to change this to be metrician has a default config (specified as yaml) and a commented out version of that our users can copy into their projects and change as they see fit, which enables partial configs and gives us a way to separate our defaults from our intentions slightly

jason-o-matic committed:
Add example config people can copy to their projects

This will make it easier for people to change specific config values for
their needs. In the docs we'll tell people to copy this to their
project and modify as needed.

This also renames the default config file to be a little more obvious.

jason-o-matic committed:
Look for configs that use yml instead of yaml

An incidental github search made it clear yml is still very popular.

jason-o-matic commented:
@janxious things have been pushed, see what you think now!

janxious commented on metrician.example.yaml:
```diff
@@ -0,0 +1,136 @@
+#   # report a metric per exception type raised
+#   :exception_specific:
+#     :enabled: false
+#
```
we might leave spaces between the sections with no comments, to indicate sections.

janxious commented on spec/metrician_spec.rb:
```diff
@@ -10,12 +10,32 @@ def self.null_agent(agent_class: Instrumental::Agent)
 RSpec.describe Metrician do
   before(:each) do
     Metrician.reset
+    ENV["METRICIAN_CONFIG"] = nil
```
More properly this should probably be `ENV.delete("METRICIAN_CONFIG")`

jason-o-matic replied:
I like this change because the intent is more clear, but something tickled my brain about it so I looked it up.

From https://ruby-doc.org/core-1.9.3/ENV.html regarding `delete`:
> Sets the environment variable name to value. If the value given is nil the environment variable is deleted.

So the functionality is actually the same!

janxious replied:
They're weirding up hashes!

janxious commented on lib/metrician/configuration.rb:
```diff
@@ -7,29 +7,49 @@ class Configuration
 
-      if File.exist?(app_location)
-        return YAML.load_file(app_location)
+    def self.merge(a_config, b_config)
```
Is this based on http://api.rubyonrails.org/classes/Hash.html#method-i-deep_merge-21 by chance?

jason-o-matic replied:
I just did it based on memory since I've written a few times in the past. I'm not tied to the implementation details if you have suggestions, though.

janxious replied:
Mostly active support is well-trod, so directly copying their code is usually where I go when I need one of their funcs. :)

janxious reviewed (approved):
some questions outstanding, but seems reasonable. Also I still think three way merge is weird.

jason-o-matic committed:
Clarify example config sections

Based on PR feedback remove blank comment lines.

jason-o-matic committed:
Clarify ENV unsetting in tests

Based on PR feedback use `delete` on ENV as it's more obvious.

jason-o-matic committed:
Use a more standard deep_merge implementation

Per PR feedback about battle tested-ness.

jason-o-matic committed:
Fix travis mysql issue

We're getting this error for all mysql configs now:
```
Mysql2::Error:
  Access denied for user 'travis'@'%' to database 'metrician_test'
```

These changes switch to use the mysql root user which I believe to be
the real fix. It also updated the travis config to use `before_install`
which is what the travis docs indicate as opposed to `before_script`,
though I think both should work.

This is based on:
https://docs.travis-ci.com/user/database-setup/#MySQL

jason-o-matic committed:
Merge remote-tracking branch 'origin/master' into fix_config

janxious reviewed (approved):
still 👍
